block-storage/src/main/scala/coop/rchain/blockstorage/InMemBlockStore.scala

```diff
@@ -0,0 +1,92 @@
+package coop.rchain.blockstorage
+
+import cats._
+import cats.implicits._
+import cats.effect.{Bracket, ExitCase}
+import coop.rchain.blockstorage.BlockStore.BlockHash
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.metrics.Metrics
+import coop.rchain.shared.SyncVarOps
+
+import scala.concurrent.SyncVar
+import scala.language.higherKinds
+
+class InMemBlockStore[F[_]] private ()(implicit
+                                       bracketF: Bracket[F, Exception],
+                                       metricsF: Metrics[F])
+    extends BlockStore[F] {
+
+  implicit val applicative: Applicative[F] = bracketF
+
+  protected[this] val stateRef: SyncVar[Map[BlockHash, BlockMessage]] =
+    SyncVarOps.create[Map[BlockHash, BlockMessage]](Map.empty)
+
+  def put(blockHash: BlockHash, blockMessage: BlockMessage): F[Unit] =
+    for {
+      _ <- metricsF.incrementCounter("block-store-put")
+      ret <- bracketF.bracket(applicative.pure(stateRef.take()))(state =>
```

Comment:
If you're already lifting actions on  the store to the `F` why not use the `MVar` or `Ref` from `cats`?

Reply:
afaik - bracket allows us to model a transaction (LMDB) without adding it's own semantics. the goal of this change is to introduce LMDB and the current PR is just one (large) step.

Reply:
From cats docs:

> a generalized abstracted pattern of safe resource acquisition and release in the face of errors or interruption.

I don't know if it can  be used instead of locks (as transaction).

Reply:
I agree that using `Ref` here can eliminate the bracket/applicative/SyncVar combo quite nicely.